### ISSUE TITLE
dict(zip(...)) confuses type checker

### ISSUE THREAD

**Issue by gjcarneiro** [python/mypy#2422]
This is checked correctly by mypy:
```python
from typing import Union, List, Dict

def f() -> Dict[str, Union[str, List[str]]]:
        return {'a', 'aa', 'b', None}
```
```
test2.py:5: error: Incompatible return value type (got Set[str], expected Dict[str, Union[str, List[str]]])
```

But this code also returns the same value but mypy gives no error:
```python
from typing import Union, List, Dict

def f() -> Dict[str, Union[str, List[str]]]:
        return dict(zip(['a', 'b'], ['aa', None]))
```
This test case exhibits similar problem.  It seems that dict(zip(...)) confuses type checking:
```python
from typing import Union, List, Dict

def f(x: Dict[str, Union[str, List[str]]]) -> None:
	pass


f(dict(zip(['a', 'b'], ['aa', None])))

f({'a', 'aa', 'b', None})
```

**Comment by gvanrossum** [comment]
I think you have a typo in your first snippet. That's a set literal, not a dict literal, and that's what the error is about. If I change the last line to
```
return {'a': 'aa', 'b': None}
```
then it also doesn't give an error. And the reason is that None is allowed as a member of any type. However, if you pass `--strict-optional` mypy will diagnose the error.

**Comment by gjcarneiro** [comment]
Ah, sorry for the waste of time, I completely forgot about `--strict-optional`.  Just trying to report a problem a colleague reported, my own code is checked with `--strict-optional`.
